File: Backend/demo/src/main/java/com/example/demo/service/GrupoService.java
package com.example.demo.service;

import com.example.demo.dto.AlumnoResponseDTO;
import com.example.demo.dto.GrupoDTO;
import com.example.demo.dto.GrupoResponseDTO;
import com.example.demo.model.Grupos;

import java.util.List;

public interface GrupoService {

    Grupos crearGrupo (GrupoDTO dto);

    List<GrupoResponseDTO> listarGrupo();

    GrupoResponseDTO obtenerGrupoPorId(Long id);

    void asignarAlumno(Long grupoId, Long alumnoId);

    void quitarAlumno(Long grupoId, Long alumnoID);

    List<AlumnoResponseDTO> listarAlumnosDelGrupo(Long grupoId);

    void eliminarGrupo(Long id);
}

File: Frontend/app/src/app/services/grupo/grupo.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { GrupoDTO, GrupoResponse } from '../../models/grupo';
import { AlumnoResponse } from '../../models/alumno.model';

@Injectable({
  providedIn: 'root'
})
export class GrupoService {

  private apiUrl = 'http://localhost:8080/api/grupos';


  constructor(private http: HttpClient) { }

  listarGrupos(): Observable<GrupoResponse[]> {
    return this.http.get<GrupoResponse[]>(this.apiUrl);
  }


  obtenerGrupoPorId(id: number): Observable<GrupoResponse> {
    return this.http.get<GrupoResponse>(`${this.apiUrl}/${id}`);
  }

  asignarAlumno(grupoId: number, alumnoId: number): Observable<void> {
    return this.http.post<void>(
      `${this.apiUrl}/${grupoId}/alumnos/${alumnoId}`,
      {}
    );
  }


  quitarAlumno(grupoId: number, alumnoId: number): Observable<any> {
    return this.http.delete(
      `${this.apiUrl}/${grupoId}/alumnos/${alumnoId}`,
      { observe: 'response'}
    );
  }


  listarAlumnosDelGrupo(grupoId: number): Observable<AlumnoResponse[]> {
    return this.http.get<AlumnoResponse[]>(
      `${this.apiUrl}/${grupoId}/alumnos`
    );
  }

  crearGrupo(grupo: GrupoDTO) {
    return this.http.post<GrupoResponse>(this.apiUrl, grupo);
  }

  eliminarGrupo(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

}
